repair/row_level.cc

```diff
@@ -2990,3 +3046,33 @@ future<> repair_service::stop() {
 repair_service::~repair_service() {
     assert(_stopped);
 }
+
+future<std::optional<gc_clock::time_point>>
+repair_service::update_history(utils::UUID repair_id, utils::UUID table_id, dht::token_range range, gc_clock::time_point repair_time) {
+    auto shard = unsigned(repair_id.get_most_significant_bits()) % smp::count;
+    return container().invoke_on(shard, [repair_id, table_id, range, repair_time] (repair_service& rs) mutable -> future<std::optional<gc_clock::time_point>> {
+        repair_history& rh = rs._finished_ranges_history[repair_id];
+        if (rh.repair_time > repair_time) {
+            rh.repair_time = repair_time;
+        }
+        auto finished_shards = ++(rh.finished_ranges[table_id][range]);
+        if (finished_shards == smp::count) {
+            // All shards have finisehd repair the range. Send an rpc to ask peers to update system.repair_history table
+            rlogger.debug("repair[{}]: Finished range {} for table {} on all shards, updating system.repair_history table, finished_shards={}",
+                    repair_id, range, table_id, finished_shards);
+            co_return rh.repair_time;
+        } else {
+            rlogger.debug("repair[{}]: Finished range {} for table {} on all shards, updating system.repair_historytable, finished_shards={}",
+                    repair_id, range, table_id, finished_shards);
+            co_return std::nullopt;
+        }
+    });
+}
+
+future<> repair_service::cleanup_history(utils::UUID repair_id) {
+    auto shard = unsigned(repair_id.get_most_significant_bits()) % smp::count;
```

Comment:
The computation for shard here and in update_history() should be the same for correctness, so it makes sense to extract it into a common function.

Reply:
OK.